week10/ns/ns.go

```diff
@@ -0,0 +1,95 @@
+package ns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+)
+
+type Ns struct {
+	cancel  context.CancelFunc
+	errs    chan error
+	subs    chan *Subscriber
+	Stopped bool
+	Name    string
+	Age     int
+	sync.RWMutex
+	sync.Once
+}
+
+const Filename = "backup.json"
+
+func (n *Ns) Start(ctx context.Context) (context.Context, error) {
+	return ctx, nil
+}
+
+func (n *Ns) Save() error {
+
+	state, err := json.Marshal(n)
```

Comment:
Look at the `json.Encoder` and `json.Decoder` types for working with `io.Writer` and `io.Reader` implementations, instead of using `os.Writefile` to directly write to disk. In your tests you can use a `bytes.Buffer` in production you can use `os.Stdout`, or `os.Create`, as an `io.Writer` for example.